{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘Biostrings’ was built under R version 4.1.3”\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Warning message:\n",
      "“package ‘S4Vectors’ was built under R version 4.1.3”\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.1.3”\n",
      "Loading required package: XVector\n",
      "\n",
      "Warning message:\n",
      "“package ‘XVector’ was built under R version 4.1.3”\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Warning message:\n",
      "“package ‘GenomeInfoDb’ was built under R version 4.1.3”\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘ape’ was built under R version 4.1.3”\n",
      "\n",
      "Attaching package: ‘ape’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biostrings’:\n",
      "\n",
      "    complement\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘ggplot2’ was built under R version 4.1.3”\n",
      "Registered S3 method overwritten by 'ggtree':\n",
      "  method      from \n",
      "  identify.gg ggfun\n",
      "\n",
      "ggtree v3.2.1  For help: https://yulab-smu.top/treedata-book/\n",
      "\n",
      "If you use ggtree in published research, please cite the most appropriate paper(s):\n",
      "\n",
      "1. Guangchuang Yu. Using ggtree to visualize data on tree-like structures. Current Protocols in Bioinformatics. 2020, 69:e96. doi:10.1002/cpbi.96\n",
      "2. Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution. 2018, 35(12):3041-3043. doi:10.1093/molbev/msy194\n",
      "3. Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ggtree’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ape’:\n",
      "\n",
      "    rotate\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biostrings’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“package ‘gridExtra’ was built under R version 4.1.3”\n",
      "\n",
      "Attaching package: ‘gridExtra’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "Registered S3 methods overwritten by 'ggalt':\n",
      "  method                  from   \n",
      "  grid.draw.absoluteGrob  ggplot2\n",
      "  grobHeight.absoluteGrob ggplot2\n",
      "  grobWidth.absoluteGrob  ggplot2\n",
      "  grobX.absoluteGrob      ggplot2\n",
      "  grobY.absoluteGrob      ggplot2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mode = ifelse(length(commandArgs(trailingOnly=TRUE))==1, 'debug', 'batch')\n",
    "\n",
    "library(Biostrings)\n",
    "library(ape)\n",
    "library(ggplot2)\n",
    "library(ggtree)\n",
    "library(gridExtra)\n",
    "library(rkftools)\n",
    "library(ggmsa)\n",
    "\n",
    "options(stringsAsFactors=FALSE)\n",
    "#options(warn=-1)\n",
    "options(repr.matrix.max.cols=200)\n",
    "#options(repr.matrix.max.rows=600, repr.matrix.max.cols=200)\n",
    "\n",
    "font_size_factor = 0.352777778"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "arguments:\n",
      "dir_csubst = /Volumes/kfT7/Dropbox/collaborators/Toshiya_Ando/20230401_insect_coloration/HOG0005974_cuticle_protein \n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (mode==\"debug\") {\n",
    "    dir_csubst = '/Volumes/kfT7/Dropbox/collaborators/Toshiya_Ando/20230401_insect_coloration/HOG0005974_cuticle_protein'\n",
    "    #dir_csubst = getwd()\n",
    "    setwd(dir_csubst)\n",
    "    args = c()\n",
    "    args = c(args, paste0('--dir_csubst=', dir_csubst))\n",
    "} else if (mode==\"batch\") {\n",
    "    args = commandArgs(trailingOnly=TRUE)\n",
    "}\n",
    "\n",
    "cat('arguments:\\n')\n",
    "args = rkftools::get_parsed_args(args, print=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_numerical_node_labels = function(phylo, rename_node=FALSE) {\n",
    "    if (rename_node) {\n",
    "        phylo[['node.label']] = paste0('n', 1:length(tree[['node.label']]))\n",
    "    }\n",
    "    all_leaf_names = phylo[['tip.label']]\n",
    "    all_leaf_names = sort(all_leaf_names)\n",
    "    leaf_numerical_labels = list()\n",
    "    for (i in 0:length(all_leaf_names)) {\n",
    "        leaf_numerical_labels[all_leaf_names[i]] = 2**i\n",
    "    }\n",
    "    intnode_numerical_labels = list()\n",
    "    for (st in subtrees(phylo)) {\n",
    "        leaf_names = st[['tip.label']]\n",
    "        numerical_label = sum(unlist(leaf_numerical_labels[leaf_names]))\n",
    "        intnode_numerical_labels[[st[['node.label']][1]]] = numerical_label\n",
    "    }\n",
    "    numerical_labels = c(unlist(leaf_numerical_labels), unlist(intnode_numerical_labels))\n",
    "    numerical_labels = sort(numerical_labels)\n",
    "    names = names(numerical_labels)\n",
    "    numerical_labels = (1:length(numerical_labels)) - 1\n",
    "    names(numerical_labels) = names\n",
    "    attr(phylo, 'numerical_label') = c(numerical_labels[phylo[['tip.label']]], numerical_labels[phylo[['node.label']]])\n",
    "    return(phylo)\n",
    "}\n",
    "\n",
    "get_line_coordinate = function(g, numerical_labels, jitter=FALSE, pairwise=FALSE) {\n",
    "    if (pairwise) {\n",
    "        nl_combinations = combn(numerical_labels, 2)\n",
    "        dat = data.frame(nl_start=nl_combinations[1,], nl_end=nl_combinations[2,])\n",
    "        for (attr in c('start','end')) {\n",
    "            dat = merge(dat, g$data[c('numerical_label','branch','y')], all.x=TRUE, by.x=paste0('nl_',attr), by.y='numerical_label')\n",
    "            colnames(dat) = sub('^branch$', paste0('x_',attr), colnames(dat))\n",
    "            colnames(dat) = sub('^y$', paste0('y_',attr), colnames(dat))\n",
    "        }\n",
    "        out = dat\n",
    "    } else {\n",
    "        is_target = (g[['data']][['numerical_label']] %in% numerical_labels)\n",
    "        dat = g[['data']][is_target,]\n",
    "        out = dat[,c('numerical_label','branch','y')]\n",
    "        colnames(out) = sub('^y$', 'y_start', colnames(out))\n",
    "        rownames(out) = NULL\n",
    "        out = out[order(out[['y_start']], decreasing=TRUE),]\n",
    "        out[,'x_start'] = out[['branch']]\n",
    "        if (jitter) {\n",
    "            #amount = \n",
    "            out[,'x_start'] = jitter(out[['x_start']], amount=amount)\n",
    "        }\n",
    "        out[,'x_end'] = out[(1:nrow(out))+1,'x_start']\n",
    "        out[,'y_end'] = out[(1:nrow(out))+1,'y_start']\n",
    "        out = na.omit(out)\n",
    "        rownames(out) = NULL\n",
    "    }\n",
    "    return(out)\n",
    "}\n",
    "\n",
    "density_scatter = function(x1,x2,\n",
    "                               ylim=c(min(x2),max(x2)),\n",
    "                               xlim=c(min(x1),max(x1)),\n",
    "                               xlab=\"\",ylab=\"\",main=\"\") {\n",
    "    # http://knowledge-forlife.com/r-color-scatterplot-points-density/\n",
    "    df <- data.frame(x1,x2)\n",
    "    x <- densCols(x1,x2, colramp=colorRampPalette(c(\"black\", \"white\")))\n",
    "    df$dens <- col2rgb(x)[1,] + 1L\n",
    "    cols <-  colorRampPalette(c(\"#000099\", \"#00FEFF\", \"#45FE4F\",\"#FCFF00\", \"#FF9400\", \"#FF3100\"))(256)\n",
    "    df$col <- cols[df$dens]\n",
    "    plot(x2~x1, data=df[order(df$dens),], \n",
    "         ylim=ylim,xlim=xlim,pch=20,col=col,\n",
    "         cex=2,xlab=xlab,ylab=ylab, las=1,\n",
    "         main=main)\n",
    "}\n",
    "\n",
    "get_cb = function(dir_csubst, arity, subsample=FALSE) {\n",
    "    csubst_files = list.files(dir_csubst)\n",
    "    if (subsample) {\n",
    "        kwd = 'csubst_cb_subsample_[0-9].*'\n",
    "    } else {\n",
    "        kwd = 'csubst_cb_[0-9].*'\n",
    "    }\n",
    "    cb_files = csubst_files[grep(kwd, csubst_files)]\n",
    "    file_path = paste0(dir_csubst, cb_files[grep(as.character(arity), cb_files)])\n",
    "    cb = read.table(file_path, sep='\\t', header=TRUE)\n",
    "    return(cb)\n",
    "}\n",
    "\n",
    "overlay_convergence = function(g, cb, stat='OCNany2spe_dev', top_percent_to_show, max_num_to_show, is_target_only, show_label=TRUE) {\n",
    "    cb[,paste0(stat, '_rank')] = rank(-cb[[stat]], na.last=TRUE, ties.method=\"first\")\n",
    "    bid_cols = colnames(cb)[grep('branch_id_', colnames(cb))]\n",
    "    arity = length(bid_cols)\n",
    "    if (is_target_only) {\n",
    "        cb_target = cb[(cb[,'branch_num_fg']==arity),]\n",
    "    } else {\n",
    "        cb_target = cb\n",
    "    }\n",
    "    num_show = floor(nrow(cb_target) * top_percent_to_show * 0.01)\n",
    "    num_show = max(num_show, 1)\n",
    "    cat(paste0(top_percent_to_show, '% (', num_show, '+tiers/', nrow(cb), ') of top protein convergence will be analyzed.\\n'))\n",
    "    cb_target = cb_target[rev(order(cb_target[,stat])),]\n",
    "    cb_target = cb_target[(cb_target[,stat]>0),]\n",
    "    rownames(cb_target) = NULL    \n",
    "    cat('There are', nrow(cb_target), 'branch combinations that satisfy', stat, '> 0\\n')\n",
    "    threshold = cb_target[num_show,stat]\n",
    "    cb_target = cb_target[(cb_target[,stat]>=threshold),]\n",
    "    if ((nrow(cb_target)>max_num_to_show)&(!is.infinite(threshold))) {\n",
    "        cat('For visualization, top protein convergence will be analyzed if branch combinations have', stat, 'equal to the value in', max_num_to_show, 'th ranked one.\\n')\n",
    "        threshold = cb_target[max_num_to_show,stat]\n",
    "        if (threshold==Inf) {\n",
    "            threshold = max(cb_target[(cb_target[,stat]!=Inf),stat])\n",
    "        }\n",
    "        cb_target = cb_target[(cb_target[,stat]>threshold),]\n",
    "    }\n",
    "    cat('There are', nrow(cb_target), 'branch combinations that satisfy', stat, '>', threshold, '\\n')\n",
    "    counts = c()\n",
    "    if (nrow(cb_target)>0) {\n",
    "        df_line_coords = data.frame()\n",
    "        for (i in 1:nrow(cb_target)) {\n",
    "            numerical_labels = unlist(cb_target[i,bid_cols])\n",
    "            line_color = ifelse(cb_target[i,'is_fg']=='Y', 'firebrick', 'gray50')\n",
    "            line_coords = get_line_coordinate(g, numerical_labels)\n",
    "            line_coords[1,'OCNany2spe'] = round(cb_target[i,'OCNany2spe'], digits=1)\n",
    "            line_coords[1,stat] = round(cb_target[i,stat], digits=1)\n",
    "            line_coords[1,paste0(stat, '_rank')] = cb_target[i,paste0(stat, '_rank')]\n",
    "            g = g + geom_curve(aes(x=x_start, y=y_start, xend=x_end, yend=y_end, alpha=0.1), \n",
    "                                                   size=0.3, curvature=jitter(0,amount=0.05), colour=line_color, \n",
    "                                                   data=line_coords, show.legend=FALSE)\n",
    "            counts = c(counts, unname(numerical_labels))\n",
    "            df_line_coords = rbind(df_line_coords, line_coords)\n",
    "        }\n",
    "        labels = paste0(df_line_coords[,'OCNany2spe'], '/', df_line_coords[,stat], '/', df_line_coords[,paste0(stat, '_rank')])\n",
    "        cat(paste0('Branch annotations: OCNany2spe/', stat, '/rank\\n'))\n",
    "        df_label_coords = data.frame(\n",
    "            'x' = df_line_coords[,'x_start'] + ((df_line_coords[,'x_end'] - df_line_coords[,'x_start'])/2),\n",
    "            'y' = df_line_coords[,'y_start'] + ((df_line_coords[,'y_end'] - df_line_coords[,'y_start'])/2),\n",
    "            label = labels\n",
    "        )\n",
    "        if (show_label) {\n",
    "            g = g + ggrepel::geom_text_repel(mapping=aes(x=x, y=y, label=label), data=df_label_coords,\n",
    "                                                                 color='firebrick', size=2, hjust=0.5, vjust=0.5, force=0.02)\n",
    "        }\n",
    "    }\n",
    "    counts = table(counts)\n",
    "    circle_coords = data.frame(numerical_label=as.numeric(names(counts)))\n",
    "    circle_coords[,'count'] = as.numeric(unname(counts))\n",
    "    g[['data']] = merge(g[['data']], circle_coords, by='numerical_label', sort=FALSE, all.x=TRUE)\n",
    "    g[['data']][(is.na(g[['data']][,'count'])),'count'] = 0\n",
    "    g[['data']][,'show_circle'] = (g[['data']][,'count']>0)\n",
    "    #g[['data']][,'is_fg'] = factor(g[['data']][['is_fg']], levels=rev(unique(g[['data']][['is_fg']])))\n",
    "    g = g + geom_point2(aes(x=branch, y=y, subset=show_circle, color=!is_fg), shape=16, size=2.5, show.legend=FALSE)\n",
    "    g = g + geom_text2(aes(x=branch, y=y, label=count, subset=show_circle), color='white', size=2, show.legend=FALSE)\n",
    "    g = g + guides(alpha=FALSE)\n",
    "    return(g)\n",
    "}\n",
    "\n",
    "extract_numerical_node_labels = function(tree) {\n",
    "    labels = c(tree[['tip.label']], tree[['node.label']])\n",
    "    numerical_labels = as.integer(sub('.*\\\\|([0-9]*)$', '\\\\1', labels))\n",
    "    numerical_labels[is.na(numerical_labels)] = -1\n",
    "    label_names = sub('(.*)\\\\|[0-9]*$', '\\\\1', labels)\n",
    "    tip_names = label_names[1:length(tree[['tip.label']])]\n",
    "    node_names = label_names[(length(tree[['tip.label']])+1):length(label_names)]\n",
    "    tree[['tip.label']] = tip_names\n",
    "    tree[['node.label']] = node_names\n",
    "    attr(tree, 'numerical_label') = numerical_labels\n",
    "    return(tree)\n",
    "}\n",
    "\n",
    "annotate_fg_tip = function(df=g$data, fg_file=file.path(dir_csubst, 'foreground.txt')) {\n",
    "    df_fg = read.table(fg_file, sep='\\t', header=FALSE)\n",
    "    colnames(df_fg) = c('lineage_num', 'regex')\n",
    "    df[,'is_fg_tip'] = FALSE\n",
    "    for (regex in df_fg[['regex']]) {\n",
    "        is_target_tip = grepl(regex, df[['label']])\n",
    "        df[is_target_tip,'is_fg_tip'] = TRUE\n",
    "    }\n",
    "    return(df)\n",
    "}\n",
    "\n",
    "get_alignment_data = function(dir_csubst, b, alignment_min_substitution) {\n",
    "    fg_subs = unlist(sapply(b[(b[['is_fg']]=='yes'),][['N_sitewise']], function(x){strsplit(x, ',')}))\n",
    "    names(fg_subs) = NULL\n",
    "    fg_subs = gsub(\"^.{1}\", \"\", fg_subs)\n",
    "    df_fg_subs = data.frame(table(fg_subs))\n",
    "    df_fg_subs[['site']] = as.integer(gsub(\".{1}$\", \"\", df_fg_subs[['fg_subs']]))\n",
    "    df_fg_subs_show = df_fg_subs[(df_fg_subs[['Freq']]>=alignment_min_substitution),]\n",
    "    aln_path = file.path(dir_csubst, 'csubst_alignment_aa.fa')\n",
    "    aln = Biostrings::readAAStringSet(aln_path)\n",
    "    data2 = tidy_msa(aln)\n",
    "    data2[['name']] = sub('\\\\|.*', '', as.character(data2[['name']]))\n",
    "    is_site = (data2[['position']] %in% df_fg_subs_show[['site']])\n",
    "    is_tip = !grepl('^n[0-9]+$', as.character(data2[['name']]))\n",
    "    data2 = data2[(is_site&is_tip),]\n",
    "    data2[['original_position']] = data2[['position']]\n",
    "    site_counter = 0\n",
    "    for (position in unique(data2[['position']])) {\n",
    "        site_counter = site_counter + 1\n",
    "        data2[(data2[['original_position']]==position),'position'] = site_counter\n",
    "    }\n",
    "    rownames(data2) = NULL\n",
    "    return(data2)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "csubst_files = list.files(args[['dir_csubst']])\n",
    "cb_files = csubst_files[grep('csubst_cb_[0-9].tsv', csubst_files)]\n",
    "cb_stat_file = csubst_files[grep('csubst_cb_stats.*', csubst_files)]\n",
    "\n",
    "tree = read.tree(file.path(args[['dir_csubst']], 'csubst_tree.nwk'))\n",
    "#tree = add_numerical_node_labels(tree)\n",
    "tree = extract_numerical_node_labels(tree)\n",
    "b = read.table(file.path(dir_csubst, 'csubst_b.tsv'), sep='\\t', header=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Arity = 3 \n",
      "no data for arity = 3 fg = N \n",
      "100% (5+tiers/5) of top protein convergence will be analyzed.\n",
      "There are 5 branch combinations that satisfy omegaCany2spe > 0\n",
      "There are 5 branch combinations that satisfy omegaCany2spe > 282.4563 \n",
      "Branch annotations: OCNany2spe/omegaCany2spe/rank\n",
      "Visualized amino acid sites: 173, 172, 136, 181, 134, 182, 85, 131, 130, 121, 145, 201, 180, 147, 81, 169, 139, 212, 92, 82, 143, 120, 237, 79, 195, 128, 88, 243, 94, 119, 90, 91, 123, 87, 186 \n",
      "\n",
      "Arity = 4 \n",
      "no data for arity = 4 fg = N \n",
      "100% (1+tiers/1) of top protein convergence will be analyzed.\n",
      "There are 1 branch combinations that satisfy omegaCany2spe > 0\n",
      "There are 1 branch combinations that satisfy omegaCany2spe > 899860.4 \n",
      "Branch annotations: OCNany2spe/omegaCany2spe/rank\n",
      "Visualized amino acid sites: 181, 195, 92, 186, 182, 147, 169, 128, 130, 134, 88 \n",
      "\n",
      "Arity = 5 \n",
      "no data for arity = 5 fg = Y \n",
      "no data for arity = 5 fg = N \n",
      "100% (1+tiers/0) of top protein convergence will be analyzed.\n",
      "There are 0 branch combinations that satisfy omegaCany2spe > 0\n",
      "There are 0 branch combinations that satisfy omegaCany2spe > NA \n",
      "Visualized amino acid sites: 169, 134, 130 \n",
      "\n",
      "Arity = 6 \n",
      "no data for arity = 6 fg = Y \n",
      "no data for arity = 6 fg = N \n",
      "100% (1+tiers/0) of top protein convergence will be analyzed.\n",
      "There are 0 branch combinations that satisfy omegaCany2spe > 0\n",
      "There are 0 branch combinations that satisfy omegaCany2spe > NA \n",
      "Visualized amino acid sites: 169 \n",
      "\n",
      "Arity = 7 \n",
      "Arity = 8 \n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "options(warn=-1)\n",
    "font_size = 8\n",
    "branch_colors = c('firebrick', 'gray50')\n",
    "min_OCNCoD = 0\n",
    "min_OCNany2any = 0\n",
    "min_OCSany2any = 0\n",
    "min_OCNany2spe = 1.5\n",
    "min_OCSany2spe = 0\n",
    "min_omegaCany2spe = 3.0\n",
    "arity_min = 3\n",
    "stat='omegaCany2spe'\n",
    "arity_max = 8\n",
    "foreground_only = TRUE\n",
    "\n",
    "df_sub = data.frame()\n",
    "for (arity in seq(arity_min, arity_max)) {\n",
    "    cat('Arity =', arity, '\\n')\n",
    "    file_path = file.path(dir_csubst, paste0('csubst_cb_', arity, '.tsv'))\n",
    "    if (!file.exists(file_path)) {\n",
    "        next\n",
    "    }\n",
    "    cb = read.table(file_path, sep='\\t', header=TRUE)\n",
    "\n",
    "    cb2 = cb\n",
    "    if (foreground_only) {\n",
    "        cb2 = cb2[(cb2[['is_fg']]=='Y'),]\n",
    "    }\n",
    "    conditions = (cb2[['OCNCoD']]>=min_OCNCoD)\n",
    "    conditions[is.na(conditions)] = FALSE\n",
    "    cb2 = cb2[(conditions),]\n",
    "    conditions = (cb2[['OCNany2spe']]>=min_OCNany2spe)\n",
    "    conditions[is.na(conditions)] = FALSE\n",
    "    cb2 = cb2[(conditions),]\n",
    "    conditions = (cb2[['OCSany2spe']]>=min_OCSany2spe)\n",
    "    conditions[is.na(conditions)] = FALSE\n",
    "    cb2 = cb2[(conditions),]\n",
    "    conditions = (cb2[['omegaCany2spe']]>=min_omegaCany2spe)\n",
    "    conditions[is.na(conditions)] = FALSE\n",
    "    cb2 = cb2[(conditions),]\n",
    "\n",
    "    for (fg in c('Y','N')) {\n",
    "        is_it = (cb2[['is_fg']]==fg)\n",
    "        if (sum(is_it)==0) {\n",
    "            cat(paste('no data for arity =', arity, 'fg =', fg, '\\n'))\n",
    "            next\n",
    "        }\n",
    "        for (col in c('OCNany2spe','OCSany2spe','OCNany2dif','dSCany2spe','dNCany2spe')) {\n",
    "            values = cb2[is_it,col]\n",
    "            tmp = data.frame(\n",
    "                'k'=arity,\n",
    "                'fg'=fg,\n",
    "                'stat'=col,\n",
    "                'value'=values,\n",
    "                stringsAsFactors=FALSE\n",
    "            )\n",
    "            df_sub = rbind(df_sub, tmp)\n",
    "        }\n",
    "    }    \n",
    "    \n",
    "    options(repr.plot.width=8, repr.plot.height=6)\n",
    "    g = ggtree::ggtree(tree, layout='rectangular')\n",
    "    g[['data']][,'numerical_label'] = as.integer(g[['data']][['numerical_label']])\n",
    "    g$data$label = unname(sapply(g$data$label, function(x){strsplit(x, split='|', fixed=TRUE)[[1]][1]}))\n",
    "    g$data = annotate_fg_tip(df=g$data, fg_file=file.path(dir_csubst, 'foreground.txt'))\n",
    "    g$data = merge(g$data, b, by.x='numerical_label', by.y='branch_id', how='left')\n",
    "    g[['data']][,'is_fg'] = (g[['data']][,'is_fg']=='yes')\n",
    "    g[['data']][,'is_mg'] = (g[['data']][,'is_mg']=='yes')\n",
    "    if (sum(g[['data']][['is_fg']])==nrow(g[['data']])) {\n",
    "        branch_colors = c('black')\n",
    "    }\n",
    "    data_aln = get_alignment_data(dir_csubst, b, alignment_min_substitution=arity)\n",
    "    data_aln = merge(data_aln, g[['data']][,c('label','is_fg_tip')], by.x='name', by.y='label', how='left')\n",
    "    \n",
    "    #g = g + ggplot2::xlim(0, max(g$data$x)+0.5)\n",
    "    g = g + scale_color_manual(values=branch_colors)\n",
    "    g = g + theme_tree()\n",
    "    g = g + geom_treescale(x=0, y=0, offset=1, fontsize=font_size*font_size_factor, linesize=1)\n",
    "    g = g + ggtree::geom_tree(mapping=aes(color=!is_fg), show.legend=FALSE)\n",
    "    g = g + ggtree::geom_tiplab(aes(color=!is_fg_tip), size=font_size*font_size_factor, align=FALSE, linetype=\"dotted\", linesize=0.5, show.legend=FALSE)\n",
    "    g = overlay_convergence(g, cb2, stat=stat, top_percent_to_show=100, max_num_to_show=Inf, is_target_only=FALSE, show_label=FALSE)\n",
    "    g = g + xlim_tree(3)\n",
    "    g = g + geom_facet(geom=geom_msa, data=data_aln,  panel='Site', font=NULL, color=\"LETTER\")\n",
    "    g = g + geom_facet(geom=geom_text, data=data_aln, panel='Site', mapping=aes(x=position, label=character), size=font_size*font_size_factor, color='white')\n",
    "    g = g + theme(\n",
    "        strip.text=element_text(size=font_size),\n",
    "    )\n",
    "\n",
    "    visualized_sites = unique(data_aln[['original_position']])\n",
    "    cat('Visualized amino acid sites:', paste(visualized_sites, collapse=', '), '\\n')\n",
    "    num_site = length(visualized_sites)\n",
    "    facet_widths(g, widths = c(1, num_site/40))\n",
    "\n",
    "    height = length(tree[['tip.label']]) / 8\n",
    "    file_name = paste0('csubst_tree.K', arity, '.pdf')\n",
    "    ggsave(file_name, height=max(3,height), width=7.2)\n",
    "    g\n",
    "    cat('\\n')\n",
    "}\n",
    "cat('Done!\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
